### PR TITLE
Bratseth/spare capacity maintainer

### PR DESCRIPTION
**Pull request by bratseth**
This maintainer attempts to schedule mitigating node moves if we get into a state where we are missing spares. The spare metric will go to 0 (causing an alert) only if no mitigating moves can be scheduled.

@bjormel fyi

### CODE REVIEW

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainerTest.java [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                                                      Optional.of("host" + ( hostOffset + i) + ".yahoo.com"),
+                                                      new Flavor(resources),
+                                                      Optional.empty(),
+                                                      NodeType.tenant);
+                node = node.allocate(application, membership, node.resources(), Instant.now());
+                nodes.add(node);
+                nodeIndex++;
+            }
+            nodes = nodeRepository.addNodes(nodes, Agent.system);
+            for (int i = 0; i < count; i++) {
```
Comment (freva) [inline node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainerTest.java:293]: Why is this needed?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        HostCapacity hostCapacity = new HostCapacity(allNodes, nodeRepository().resourcesCalculator());
+        Set<Node> spareHosts = hostCapacity.findSpareHosts(allNodes.hosts().satisfies(node.resources()).asList(), 2);
+        List<Node> hosts = allNodes.hosts().except(spareHosts).asList();
+
+        CapacitySolver capacitySolver = new CapacitySolver(hostCapacity, maxIterations);
+        List<Move> shortestMitigation = null;
+        for (Node spareHost : spareHosts) {
+            List<Move> mitigation = capacitySolver.makeRoomFor(node, spareHost, hosts, List.of(), List.of());
+            if (mitigation == null) continue;
+            if (shortestMitigation == null || shortestMitigation.size() > mitigation.size())
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java:117]: It's not obvious to me that this return `null`/empty list always ends being correct, why not always return non-null?

Also, add a check here to ignore empty mitigation if we already have a `shortestMitigation` candidate, after all doing something that takes longer should be better than doing nothing?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeList.java [hunk 3]
```diff
@@ -133,6 +146,16 @@ public NodeList state(Collection<Node.State> nodeStates) {
         return matching(node -> nodeStates.contains(node.state()));
     }
 
+    /** Returns the subset of nodes in the active state */
+    public NodeList active() {
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeList.java:150]: Imho the only purpose of this class is to make node filtering statements shorter, so why not go all in.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MaintenanceDeployment.java [hunk 4]
```diff
@@ -128,4 +131,106 @@ public String toString() {
...
+                    log.info(agent + " redeployed " + application + " to " +
+                             ( verifyTarget ? this : "move " + (node.hostname() + " from " + fromHost)));
+                    return true;
+                }
+                finally {
+                    markWantToRetire(node, false, agent, nodeRepository); // Necessary if this failed, no-op otherwise
+
+                    // Immediately clean up if we reserved the node but could not activate or reserved a node on the wrong host
+                    expectedNewNode.flatMap(node -> nodeRepository.getNode(node.hostname(), Node.State.reserved))
+                                   .ifPresent(node -> nodeRepository.setDirty(node, Agent.Rebalancer, "Expired by " + agent));
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/MaintenanceDeployment.java:189]: Thanks!

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Agent.java [hunk 5]
```diff
@@ -21,6 +21,7 @@
     ProvisionedExpirer,
     ReservationExpirer,
     DynamicProvisioningMaintainer,
-    RetiringUpgrader;
+    RetiringUpgrader,
+    SpareCapacityMaintainer
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Agent.java:25]: Done

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/HostCapacity.java [hunk 6]
```diff
@@ -16,17 +20,38 @@
...
+    public HostCapacity(NodeList allNodes, HostResourcesCalculator hostResourcesCalculator) {
         this.allNodes = Objects.requireNonNull(allNodes, "allNodes must be non-null");
         this.hostResourcesCalculator = Objects.requireNonNull(hostResourcesCalculator, "hostResourcesCalculator must be non-null");
     }
 
-    int compareWithoutInactive(Node hostA, Node hostB) {
+    public NodeList allNodes() { return allNodes; }
+
+    /**
+     * Spare hosts are the two hosts in the system with the most free capacity.
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/HostCapacity.java:36]: It's really the two  ...

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+            return (number & (1 << position)) > 0;
+        }
+
+        private int onesIn(int number) {
+            int ones = 0;
+            for (int position = 0; Math.pow(2, position) <= number; position++) {
+                if (hasOneAtPosition(position, number))
+                    ones++;
+            }
+            return ones;
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java]: Didn't know about bitCount - thanks!

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/HostCapacity.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/HostCapacity.java:36]: As far as this method is concerned, it's the `count` hosts?

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainerTest.java (hunk 1 above)
Comment (bratseth) [reply node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainerTest.java:293]: The second loop? This whole thing is in lieu of allocating properly, because here I want full control over where nodes are placed, without being dependent on constraints applied during allocation. The second loop is needed because setting nodes ready will remove the allocation if I set it at creation.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java (hunk 2 above)
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java:117]: Empty list means the solution needs no moves.
null means there are no solutions.

I don't understand the second paragraph - if it's null we continue because we found no solution. Otherwise we compare it to the current best, if any.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeList.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeList.java:150]: Aside from polluting the interface, it also has disadvantage with code search. I often search for usage of those enum values to see which cases I might need to consider. If we end up having a method for every/most states, that just slows that process down where you would also need to search for usage of those methods.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+        private NodeResources freeCapacityWith(List<Move> moves, Node host) {
+            NodeResources resources = hostCapacity.freeCapacityOf(host);
+            for (Move move : moves) {
+                if ( ! move.toHost().equals(host)) continue;
+                resources = resources.subtract(move.node().resources());
+            }
+            for (Move move : moves) {
+                if ( ! move.fromHost().equals(host)) continue;
+                resources = resources.add(move.fromHost().resources());
```
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java:247]: 💯 thanks!

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java (hunk 2 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java:117]: If `makeRoomFor()` returns empty list, it will be set as the `shortestMitigation` since `shortestMitigation.size() > mitigation.size()` == `shortestMitigation.size() > 0`, but then 2 lines below, if `shortestMitigation.isEmpty()` we return `Move.empty()`. So wouldn't it be better to change this if to
```
if (shortestMitigation == null || (shortestMitigation.size() > mitigation.size() && !mitigation.isEmpty()))
```
?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/HostCapacity.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/HostCapacity.java:36]: Also, don't we need to set these to 2:
https://github.com/vespa-engine/vespa/blob/193c96ee9dd092d3f6c483a9d65f9a7dd8efa575/node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeRepositoryProvisioner.java#L49
?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeList.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeList.java]: Don't agree with those and the latter is really an argument against all library code, but as a compromise I removed active() since it was used once, but not hosts() since it was used twice :-)

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java (hunk 2 above)
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java:117]: If the shortest mitigation is empty it means that the situation can be mitigated in 0 moves, i.e it is already fine.
This is an inconsistency that will only arise if some change happened to mitigate it between when the nodes were read by CapacityChecker and by this, or if either the CapacityChecker or this has a bug. I don't think we should prefer an unnecessary non-empty mitigation in this case, but arguably we could to signal it upwards such that we don't emit the alert in this case (and perhaps also log it). I'm uncertain on whether it's worth the extra complication.

**Review by freva — APPROVED** [review]
LGTM, but I don't think

> Allocation will assign the two most empty nodes as "spares", which will not be allocated on
> unless needed for node failing. Our goal here is to make room on these spares for the given node

is true anymore, see the unresolved comment.

**Comment by bratseth** [comment]
Thanks, I didn't know we had turned off spares. No wonder this alert goes off all the time then.
I think we should turn it back on, but this maintainer should help in any case.
